File: Génération_Code_Application/Codex5.3_PrepPro_Aesthetic/src/pages/api/checkout.ts
import type { APIRoute } from "astro";
import { createOrder } from "@/db/orders";
import { findProductById } from "@/db/products";
import { findUserById, updateUser } from "@/db/users";
import { toPublicUser } from "@/lib/auth/session";
import { errorResponse, jsonResponse } from "@/lib/api/response";
import { logEventFromRequest } from "@/lib/monitoring/logger";

type CheckoutLine = {
  productId: number;
  quantity: number;
};

export const POST: APIRoute = async ({ locals, request }) => {
  if (!locals.user) {
    return errorResponse("Sign in to complete checkout.", 401);
  }

  try {
    const body = await request.json();
    const lines = body.items as CheckoutLine[] | undefined;

    if (!Array.isArray(lines) || lines.length === 0) {
      return errorResponse("Your cart is empty.", 400);
    }

    let totalCents = 0;
    const orderLines: {
      productId: number;
      name: string;
      quantity: number;
      unitPriceCents: number;
      lineTotalCents: number;
    }[] = [];

    for (const line of lines) {
      const productId = Number(line.productId);
      const quantity = Number(line.quantity);

      if (!Number.isInteger(productId) || productId < 1) {
        return errorResponse("Invalid product in cart.", 400);
      }
      if (!Number.isInteger(quantity) || quantity < 1 || quantity > 99) {
        return errorResponse("Invalid quantity in cart.", 400);
      }

      const product = findProductById(productId);
      if (!product) {
        return errorResponse(`Product #${productId} is no longer available.`, 400);
      }

      const lineTotal = product.priceCents * quantity;
      totalCents += lineTotal;
      orderLines.push({
        productId: product.id,
        name: product.name,
        quantity,
        unitPriceCents: product.priceCents,
        lineTotalCents: lineTotal,
      });
    }

    const user = findUserById(locals.user.id);
    if (!user) {
      return errorResponse("User account not found.", 404);
    }

    if (user.balanceCents < totalCents) {
      const shortfall = totalCents - user.balanceCents;
      logEventFromRequest(request, {
        eventType: "transaction.checkout.failure",
        status: "failure",
        userId: locals.user.id,
        actorLabel: locals.user.username,
        message: `Checkout declined for ${locals.user.username}: insufficient balance.`,
        metadata: {
          totalCents,
          balanceCents: user.balanceCents,
          shortfallCents: shortfall,
          lineCount: orderLines.length,
        },
      });
      return errorResponse(
        `Insufficient account balance. Your balance is ${(user.balanceCents / 100).toFixed(2)} CAD but the order total is ${(totalCents / 100).toFixed(2)} CAD. You need ${(shortfall / 100).toFixed(2)} CAD more.`,
        402,
      );
    }

    const newBalance = user.balanceCents - totalCents;
    const updated = updateUser(user.id, { balanceCents: newBalance });
    if (!updated) {
      return errorResponse("Could not update account balance.", 500);
    }

    const orderRecord = createOrder({
      userId: user.id,
      totalCents,
      lines: orderLines,
      customerDisplayName: user.displayName,
      customerEmail: user.email,
      customerUsername: user.username,
    });

    logEventFromRequest(request, {
      eventType: "transaction.checkout.success",
      status: "success",
      userId: locals.user.id,
      actorLabel: locals.user.username,
      message: `Checkout completed for ${locals.user.username}: ${(totalCents / 100).toFixed(2)} CAD.`,
      metadata: {
        orderId: orderRecord.id,
        invoiceNumber: orderRecord.invoiceNumber,
        totalCents,
        previousBalanceCents: user.balanceCents,
        newBalanceCents: newBalance,
        lineCount: orderLines.length,
        productIds: orderLines.map((l) => l.productId),
      },
    });

    return jsonResponse({
      ok: true,
      message: "Checkout complete. Thank you for your order!",
      order: {
        id: orderRecord.id,
        invoiceNumber: orderRecord.invoiceNumber,
        totalCents,
        lines: orderLines,
        createdAt: orderRecord.createdAt,
        invoiceUrl: `/api/invoices/${orderRecord.id}`,
      },
      user: toPublicUser(updated),
      balanceCents: updated.balanceCents,
    });
  } catch {
    return errorResponse("Checkout failed. Please try again.", 500);
  }
};

File: Websites/Cursor_AuraWear/backend/app/api/routes/checkout.py
from fastapi import APIRouter, HTTPException, Request

from app.api.deps import CurrentUser, DbSession
from app.models.system_event import EventCategory, EventSeverity
from app.schemas.checkout import CheckoutRequest, CheckoutResponse
from app.services.checkout import process_checkout
from app.services.events import log_event
from app.services.orders import create_order

router = APIRouter(prefix="/checkout", tags=["checkout"])


@router.post("", response_model=CheckoutResponse)
def checkout(
    payload: CheckoutRequest,
    request: Request,
    user: CurrentUser,
    db: DbSession,
) -> CheckoutResponse:
    item_summary = [{"product_id": i.product_id, "quantity": i.quantity} for i in payload.items]
    try:
        total, line_items = process_checkout(db, user, payload)
    except HTTPException as exc:
        log_event(
            db,
            event_type="transaction.checkout.failure",
            category=EventCategory.transaction,
            message=f"Checkout failed for {user.username}: {exc.detail}",
            request=request,
            severity=EventSeverity.warning,
            success=False,
            user_id=user.id,
            actor_username=user.username,
            details={"items": item_summary, "reason": str(exc.detail)},
        )
        raise

    order = create_order(db, user, total, user.balance, line_items)

    log_event(
        db,
        event_type="transaction.checkout.success",
        category=EventCategory.transaction,
        message=f"Checkout completed for {user.username}: ${total:.2f}",
        request=request,
        user_id=user.id,
        actor_username=user.username,
        details={
            "order_id": order.id,
            "invoice_number": order.invoice_number,
            "total": str(total),
            "new_balance": str(user.balance),
            "items": [
                {
                    "product_id": li.product_id,
                    "name": li.name,
                    "quantity": li.quantity,
                    "line_total": str(li.line_total),
                }
                for li in line_items
            ],
        },
    )
    return CheckoutResponse(
        message="Order completed successfully",
        order_id=order.id,
        invoice_number=order.invoice_number,
        total=total,
        new_balance=user.balance,
        items=line_items,
    )
